packages/opentelemetry-exporter-prometheus/src/prometheus.ts

```diff
@@ -0,0 +1,296 @@
+/*!
+ * Copyright 2019, OpenTelemetry Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { ExportResult } from '@opentelemetry/base';
+import { NoopLogger } from '@opentelemetry/core';
+import {
+  LabelValue,
+  MetricDescriptor,
+  MetricDescriptorType,
+  MetricExporter,
+  ReadableMetric,
+} from '@opentelemetry/metrics';
+import * as types from '@opentelemetry/types';
+import { createServer, IncomingMessage, Server, ServerResponse } from 'http';
+import { Counter, Gauge, labelValues, Metric, Registry } from 'prom-client';
+import * as url from 'url';
+import { ExporterConfig } from './export/types';
+
+export class PrometheusExporter implements MetricExporter {
+  static readonly DEFAULT_OPTIONS = {
+    port: 9464,
+    startServer: false,
+    endpoint: '/metrics',
+    prefix: '',
+  };
+
+  private readonly _registry = new Registry();
+  private readonly _logger: types.Logger;
+  private readonly _port: number;
+  private readonly _endpoint: string;
+  private _server: Server;
+  private readonly _prefix?: string;
+  private readonly _invalidCharacterRegex = /[^a-z0-9_]/gi;
+
+  // This will be required when histogram is implemented. Leaving here so it is not forgotten
+  // Histogram cannot have a label named 'le'
+  // private static readonly RESERVED_HISTOGRAM_LABEL = 'le';
+
+  /**
+   * Constructor
+   * @param config Exporter configuration
+   * @param callback Callback to be called after a server was started
+   */
+  constructor(config: ExporterConfig = {}, callback?: () => void) {
+    this._logger = config.logger || new NoopLogger();
+    this._port = config.port || PrometheusExporter.DEFAULT_OPTIONS.port;
+    this._prefix = config.prefix || PrometheusExporter.DEFAULT_OPTIONS.prefix;
+    this._server = createServer(this._requestHandler);
+
+    this._endpoint = (
+      config.endpoint || PrometheusExporter.DEFAULT_OPTIONS.endpoint
+    ).replace(/^([^/])/, '/$1');
+
+    if (config.startServer || PrometheusExporter.DEFAULT_OPTIONS.startServer) {
+      this.startServer(callback);
+    } else if (callback) {
+      callback();
+    }
+  }
+
+  /**
+   * Save current metric state so that it can be pulled by the metrics backend.
```

Comment:
IMO comments should be descriptive (Adds or Returns or Saves) rather than imperative (Add or Return or Save). WDYT?

Reply:
I don't really have a preference. I'll take a look at the comments in the exporter.

Reply:
https://github.com/open-telemetry/opentelemetry-js/pull/483/commits/ccd5ce434da0e4f2852b184e24067ca9babfc838